File: C/Algorithms/515.c
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     struct TreeNode *left;
 *     struct TreeNode *right;
 * };
 */
void visit(struct TreeNode *node, int level, int **maxes, int *size) {
  int value = node->val;
  if (level >= *size) {
    *size = *size + 1;
    *maxes = realloc(*maxes, (*size) * sizeof(int));
    (*maxes)[level] = value;
  } else {
    int prev_max = (*maxes)[level];
    (*maxes)[level] = prev_max > value ? prev_max : value;
  }
  struct TreeNode *left = node->left, *right = node->right;
  if (left) {
    visit(left, level + 1, maxes, size);
  }
  if (right) {
    visit(right, level + 1, maxes, size);
  }
}

/**
 * Note: The returned array must be malloced, assume caller calls free().
 */
int *largestValues(struct TreeNode *root, int *returnSize) {
  *returnSize = 0;
  if (!root) {
    return NULL;
  }
  int *maxes = malloc(sizeof(int));
  visit(root, 0, &maxes, returnSize);
  return maxes;
}

File: Rust/Algorithms/515.rs
use std::cell::RefCell;
use std::cmp;
use std::rc::Rc;

impl Solution {
    pub fn largest_values(
        root: Option<Rc<RefCell<TreeNode>>>,
    ) -> Vec<i32> {
        let mut maxes = vec![];

        if root.is_none() {
            return maxes;
        }

        let mut stack = vec![(root.unwrap(), 0)];
        while stack.len() > 0 {
            let (node, level) = stack.pop().unwrap();
            let node = Rc::try_unwrap(node).unwrap().into_inner();
            if level >= maxes.len() {
                maxes.push(node.val);
            } else {
                maxes[level] = cmp::max(maxes[level], node.val);
            }
            if let Some(left) = node.left {
                stack.push((left, level + 1));
            }
            if let Some(right) = node.right {
                stack.push((right, level + 1));
            }
        }
        maxes
    }
}
